Add article on how to upgrade to embark 4 (#141)
embark-framework/embark-site: source/_posts/2019-03-11-upgrading-to-embark-4.md

title: How to upgrade to Embark 4
summary: "In this guide, we'll show how to upgrade from an Embark 3.x project to Embark 4"
categories:
  - tutorials
---

The release of Embark 4.0 is close at hand and beta.2 is the relase to introduce some real breaking changes. Let's dive right into them.

> ## Use the pipeline that YOU want

Reviewer: ```suggestion
## Use **any** frontend build tool!
```

Thereby, we announce that you can now use the file pipeline that YOU want. Also, you can STILL use Embark's pipeline if you are fine with what is currently there.

So, as you just read, it's not really a breaking change, well  that is, unless you decide you want to convert you Dapp from Embark's pipeline to the one of your choice.

If you don't plan on starting a project with another pipeline or don't plan of converting, you can skip to [the next section](#New-Web3-plugin).

### Converting to another pipeline

The first thing you might want to do is change or add the new `generationDir` property in `embark.json`. This property tells Embark where to generation the various artifacts that you will need to build your Dapp. Most of those artifacts were already generated before, but were inside the `.embark/` folder. 

Since most file pipelines need those files inside the source folder, we're going to put `generationDir` as your selected pipeline's source folder. For example,  create-react-app (CRA) has `src/` as source folder.

Afterwards, we need to convert all "magic" imports to relative imports.

> The first one is the EmbarkJS import. The "magic" import is `"Embark/EmbarkJS"`. We need to convert that to the relative path. The `embarkjs.js` script is at the root of  `embarkArtifacts/`, so you need to convert it to something like `"./embarkArtifacts/embarkjs"`.

Reviewer: ```suggestion
The first one is the EmbarkJS import. The "magic" import is `"Embark/EmbarkJS"`. Anywhere we have `"Embark/EmbarkJS"` in our Dapp's code, we need to convert that to the relative path. Because we are trying to get the `EmbarkJS` library, and the `embarkjs.js` script is located in the root of  `embarkArtifacts/`, we need to replace 
```
```
import EmbarkJS from "Embark/EmbarkJS"
```
with
```
import EmbarkJS from "./embarkArtifacts/embarkjs"
```
NOTE: The relative path is dependent upon the `generationDir` setting specified in `embark.json` (see the "Artifact generation directory" section above](#Artifact generation directory).

Second ones are the contracts. They used to look like `"Embark/contract/ContractName"`. They are now located inside `embarkArtifactscontracts/`.

> Third one, there used to be `"Embark/web3"`, but we removed it. Don't worry, we are not removing web3 support, far from it. We actually just got rid of that useless import that didn't do much. All you have to do is use the global `web3` object that is avilable everywhere in the Dapp.

Reviewer: ```suggestion
Thirdly, there used to be `import web3 from "Embark/web3"`, but it has been removed in Embark 4 in favor of using a global Web3 object. Don't worry, Embark is not removing web3 support, far from it. We actually just got an import that did not provide a lot of benefit. In Embark 4, the global `web3` object is now available everywhere in the Dapp.
```

> Then, you can move your Embark files and configs inside a project created by the pipeline of your choice.

Reviewer: ```suggestion
Now, all the Embark files and configs from your Dapp can be moved in to a project created by the frontend build tool of your choice.
```

### New project with another pipeline

> If you are starting your project from scratch, you have two options.

Reviewer: ```suggestion
Starting a new Dapp from scratch is easy, we have two options.
```

> The easier one is to use our new Embark CRA template (TODO add template URl). It sets up a simple Embark project for you with the all of the files you are used to. The only slight difference is that the config files are in `embarkConfig` to make sure they don't clash with CRA's config. It's already set up for you to start and use. You only need to do `embark run` and in another terminal `yarn start` or `npm run start` to start CRA.

Reviewer: ```suggestion
The easiest option is to use our new Embark CRA template (TODO add template URl). It sets up a simple Embark project with all of the familiar files present in an Embark 3.x Dapp, with one minor difference: the config files are located in an `embarkConfig/` folder in the root of the Dapp to make sure they don't clash with CRA's config folder/files.
```
To get started with Embark's CRA template,
```
npx embark-react my-app
cd my-app
yarn start // or alternatively, npm run start
```
Then, in another terminal,
```
cd my-app
embark run
```
That's it!

> The other solution is to create a project with your chosen pipeline and then (TODO new command or we need to do it in another).

Reviewer: ```suggestion
The other solution is to create a project using a frontend build tool like Angular CLI, then (TODO new command or we need to do it in another).
```

## New Web3 plugin

> Starting with Embark 4 beta.2, Embark no longer comes with Web3js by default for your Dapp. Don't run. We did that so that we can now have the possibility of supporting more than just Web3js, but potentially EthersJS and more (even the possibility for you to do it).

Reviewer: ```suggestion
Starting with Embark 4 beta.2, Embark no longer supplies the Dapp with `Web3.js` by default. Don't run. We did that so that we can now have the possibility of supporting more than just `Web3.js`, such as EthersJS, and more. You can even roll your own.
```

> Anyway, to continue having Web3js inside your Dapp, you only have to call the following command in the Embark console: `plugin install embarkjs-connector-web3`

Reviewer: ```suggestion
To continue using `Web3.js` inside the Embark 4 Dapp, execute the following command in the Embark console: `plugin install embarkjs-connector-web3`.
```

> As you can see, it's just a plugin. You can install it manually instead by doing:

Reviewer: ```suggestion
This simply [installs `embarkjs-connector-web3` as a plugin](https://embark.status.im/docs/installing_plugins.html). Alternatively, this plugin can be installed manually by executing:
```

1. `yarn add embarkjs-connector-web3` or `npm install --save embarkjs-connector-web3`
2. Adding `"embarkjs-connector-web3": {}` to the `plugins` section of `embark.json`

> It's as simple as that. This plugin will add the necessary commands and code for your Dapp to connect to the blockchain and register the necessary providers. The only prerequisite is that you import `EmbarkJS` at least once. That file is located at `embarkArtifacts/embarkjs.js` or if you still use Embark's pipeline, you can import it using `import EmbarkJS from "Embark/EmbarkJS";`.

Reviewer: ```suggestion
It's as simple as that. This plugin will add the necessary commands and code for the Dapp to connect to the blockchain and register the necessary providers. The only prerequisite is for the Dapp to import `EmbarkJS` at least once. If using a third party pipeline, the `EmbarkJS` file can be imported using `import EmbarkJS from "./embarkArtifacts/embarkjs.js"` (or as specified by the `generationDir` in `embark.json`). If using Embark's built-in pipeline, `EmbarkJS` can be imported using `import EmbarkJS from "Embark/EmbarkJS";`.
```

## New Blockchain account configs

> In Embark 4 alpha, we added some new blockchain account configurations. They are really similar to the ones you have in the contract configuration. You can find more details in our [Accounts Blockchain configuration guide](https://embark-site-develop.netlify.com/docs/blockchain_accounts_configuration.html).

Reviewer: ```suggestion
Embark 4 adds some new blockchain account configurations. To try to keep things as simple as possible, these additions are really similar to the ones in the contract configuration. For more information, please read the [Accounts Blockchain configuration guide](https://embark.status.im/docs/blockchain_accounts_configuration.html) in our docs.
```

However, we did introduce some small breaking changes. We removed: 

> - `account`: This is completely replaced by the new `accounts` property. It gives you more flexibility. To have exactly the same behavior as before, just use the `nodeAccount` account type

Reviewer: ```suggestion
- `account`: This is completely replaced by the new `accounts` property (notice the `s` at the end of `accounts`). It gives the developer more flexibility. To have exactly the same behavior as before, just use the `nodeAccounts` account type as [described in the docs](https://embark.status.im/docs/blockchain_accounts_configuration.md#parameter-descriptions)
```

> - `simulatorMnemonic`: Removed because it was mostly useless. Ganache CLI has a default mnemonic. Also, you get the same behavior as before with the `mnemonic` account type

Reviewer: ```suggestion
- `simulatorMnemonic`: Removed in favor of Ganache's default mnemonic. If this functionality is still needed, please specify the desired mnemonic in the [blockchain config's `mnemonic` account type](https://embark.status.im/docs/blockchain_accounts_configuration.md#parameter-descriptions).
```

## Conclusion

> Embark 4 contains way more new features than those listed above. Once we release officially Embark 4, we'll list all of those cool features for you.

Reviewer: ```suggestion
This is a small taste of the features added to Embark 4, namely the ability to use a frontend build tool of choice. However, Embark 4 is jam-packed with additional new features, which we detail during the Embark 4 release.
```

> In the meantime, you can see that all that goodness doesn't come at too high a price it terms of breaking changes.

Reviewer: ```suggestion
In the meantime, all the Embark 4 goodness doesn't come at too high a price in terms of breaking changes.
```

> Hopefully, you'll have a blast upgrading to Embark 4 and if you ever have an issue, make sure to hit us on [Gitter](https://gitter.im/embark-framework/Lobby).

Reviewer: ```suggestion
Upgrading to Embark 4 will be a blast. If you ever have an issue, make sure to hit us up on [Gitter](https://gitter.im/embark-framework/Lobby).
```
